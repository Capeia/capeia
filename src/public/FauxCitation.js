// @flow
import React from 'react'
import { Popover, OverlayTrigger } from 'react-bootstrap'
import withStyles from 'isomorphic-style-loader/lib/withStyles'
import sd from './shared/rich-text-editor/entity/Decorator.scss'
import s from './FauxCitation.scss'
import citationIcon from 'article/glyphicon-education.png'

const IDEOGRAPHIC_SPACE = '\u3000'

type Props = {
  id: string,
  // TODO: Can these be null?
  data: {
    year: number,
    author: ?string,
    doi: ?string,
    source: ?string,
    title: ?string,
    url: ?string
  }
}

class FauxCitation extends React.Component {
  props: Props

  renderOverlay = () => {
    const { author, doi, source, title, year, url } = this.props.data

    const sourceLink = !url ? source : <a href={url} target='_blank'>{source}</a>

    return (
      <Popover id={this.props.id}>
        <table className={`${s.FauxCitation}`}>
          <tbody>
            {author &&
              <tr>
                <td>Author(s)</td>
                <td>{author}</td>
              </tr>
            }

            {title &&
              <tr>
                <td>Title</td>
                <td>{title}</td>
              </tr>
            }

            {year &&
              <tr>
                <td>Year</td>
                <td>{year}</td>
              </tr>
            }

            {source &&
              <tr>
                <td>Source</td>
                <td>{sourceLink}</td>
              </tr>
            }

            {doi &&
              <tr>
                <td>Volume / DOI</td>
                <td>{doi}</td>
              </tr>
            }
          </tbody>
        </table>
      </Popover>
    )
  }

  render () {
    return (
      <OverlayTrigger trigger='click' placement='top' overlay={this.renderOverlay()}>
        <span className={`${sd.ImageDecorator} ${s.icon}`} style={{backgroundImage: `url(${citationIcon})`}}>
          {IDEOGRAPHIC_SPACE}
        </span>
      </OverlayTrigger>
    )
  }
}

export default withStyles(s, sd)(FauxCitation)
