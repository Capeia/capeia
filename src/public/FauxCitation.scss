@import '~theme/typography';

.FauxCitation {
  tr {
    vertical-align: top !important;
  }

  td {
    padding-bottom: 0.4em;
  }

  td:first-child {
    @include sans-serif-bold;
    padding-right: 1em;
    white-space: nowrap;
  }
}

.icon {
  cursor: pointer;
}
