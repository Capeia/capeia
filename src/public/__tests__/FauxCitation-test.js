// @flow
/* eslint-env jest, jasmine */

import React from 'react'
import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'
import FauxCitation from '../FauxCitation'

describe('<FauxCitation>', () => {
  const data = {
    year: 1970,
    author: 'Jon Doe',
    doi: '123',
    source: 'Foo',
    title: 'Baz'
  }

  it('renders correctly', () => {
    jest.deepUnmock('../FauxCitation')
    const Component = require('../FauxCitation').default
    const tree = renderer.create(
      <Component id='123' data={data} />
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should render a popover containing correct table data', () => {
    const wrapper = shallow(<FauxCitation id='123' data={data} />)
    const trigger = wrapper.find('OverlayTrigger')
    expect(trigger.length).toEqual(1)
    const popover = shallow(trigger.props().overlay)
    expect(popover.contains(<tr><td>Author(s)</td><td>{data.author}</td></tr>)).toEqual(true)
    expect(popover.contains(<tr><td>Title</td><td>{data.title}</td></tr>)).toEqual(true)
    expect(popover.contains(<tr><td>Year</td><td>{data.year}</td></tr>)).toEqual(true)
    expect(popover.contains(<tr><td>Volume / DOI</td><td>{data.doi}</td></tr>)).toEqual(true)
    expect(popover.contains(<tr><td>Source</td><td>{data.source}</td></tr>)).toEqual(true)
  })
})
