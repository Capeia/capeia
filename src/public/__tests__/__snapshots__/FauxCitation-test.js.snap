// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FauxCitation> renders correctly 1`] = `
<span
  className="ImageDecorator icon"
  onClick={[Function]}
  style={
    Object {
      "backgroundImage": "url(article/glyphicon-education.png)",
    }
  }
>
  ã€€
</span>
`;
